pep-9999.rst

```diff
@@ -0,0 +1,187 @@
+PEP: 0999
+Title: Python minimization
+Version: $Revision$
+Last-Modified: $Date$
+Author: 
+Sponsor: 
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 15-Sep-2020
+
+
+Abstract
+========
+
+This PEP aims to provide a way for downstream distributors of Python, such as Linux distributions, to minimize the footprint of Python on the hard drive for constrained environments like containers. Two of the possible solutions are provided for discussion.
+
+
+Motivation
+==========
+
+The footprint of an installed Python tends to be quite large on most distributions.
+This is not always desirable. In use cases such as containers, there are efforts to keep the disk size as minimal as possible. We would like to provide a standardized way to accomplish this minimization so that downstream distributors don't have to reinvent the wheel, rely on hacks and/or break users expectations.
+
+
+Rationale
+=========
+
+The authors of this PEP have not settled on the specific mechanism for minimization yet. We would like to present two possible approaches for discussion.
+
+
+Option 1: Shipping only ``.py`` files, and disabling creation of ``.pyc`` files
+-------------------------------------------------------------------------------
+
+A simple way of saving space is not shipping ``.pyc`` bytecode files, as they are not strictly necessary. Python programs will run fine from only ``.py`` source files. They will take a bit longer to start, but in our testing the slowdown is acceptable. For example, importing ``importlib.py`` takes on average 0.025s longer on our machines compared to the ``.pyc`` bytecode file.
+
+The problem with this method is that when Python imports a ``.py`` source file without finding a corresponding and up to date ``.pyc`` bytecode file in ``__pycache__``, it will try to create it. This is undesirable for three reasons:
+
+1. when Python is run by the user, they likely won't have permissions to create the file on the root filesystem, and an error message will be generated,
```

Comment:
what error message?

Reply:
Huh, I guess I just assumed there would be, but there's none. How about I mention the SElinux denial? Or is that too RH-specific?

Reply:
it isn't RH specific, it is SELinux specific. That makes sense. But do note that SELinux denial audit message can be fixed.

Reply:
It shows no error to me when Python tries to create a new pyc file in a folder I don't have write permissions to and I cannot find any report in selinux logs 🤷‍♂️

Reply:
For SELinux logs, try https://github.com/fedora-selinux/selinux-policy/pull/404#issuecomment-685653823

Reply:
Changed to mention SELinux warnings instead. And the point was moved to the Motivation section.